Feature request: option to unsoundly ignore more errors issued because of #979

False positives for checkers that only issue a few errors (like the [Object Construction Checker](https://github.com/kelloggm/object-construction-checker)) are dominated by errors caused by #979. When running these checkers, suppressing these warnings is time-consuming and inconvenient for users, because they are unrelated to the property or properties actually being checked.

Some errors related to #979 are already [unsoundly suppressed](https://github.com/typetools/checker-framework/issues/979#issuecomment-279578221) by default. This feature request is to suggest that we do something similar for another class of errors. Whether or not the default is the conservative behavior isn't important to me.

The errors arise from programs like the following:

```java
import java.lang.Class;
import com.sun.tools.javac.code.Symbol;
import com.sun.tools.javac.code.Symbol.ClassSymbol;
import javax.lang.model.element.ElementKind;
import com.sun.tools.javac.code.Scope;
import com.sun.tools.javac.code.Type;
import com.sun.tools.javac.code.Types;
import com.sun.tools.javac.util.Name;

class Issue979RemainingError {
    private final Types types = null;

    private <T extends Symbol> T getMember(
      Class<T> type, ElementKind kind, Symbol classSymbol, Name name) {
	if (classSymbol.type == null) {
	    return null;
	}
	for (Type t : types.closure(classSymbol.type)) {
	    Scope scope = t.tsym.members();
	    for (Symbol sym : scope.getSymbolsByName(name)) {
		if (sym.getKind().equals(kind)) {
		    return type.cast(sym);
		}
	    }
	}
	if (classSymbol.hasOuterInstance()) {
	    T sym = getMember(type, kind, classSymbol.type.getEnclosingType().asElement(), name);
	    if (sym != null) {
		return sym;
	    }
	}
	if (classSymbol.owner != null
	    && classSymbol != classSymbol.owner
	    && classSymbol.owner instanceof Symbol.ClassSymbol) {
	    T sym = getMember(type, kind, classSymbol.owner, name);
	    if (sym != null && sym.isStatic()) {
		return sym;
	    }
	}
	return null;
    }
}
```

Typechecking this class with e.g. the Value Checker, which should issue no errors, does issue two errors.

The command I ran:
```
javac -processor org.checkerframework.common.value.ValueChecker Issue979RemainingErrors.java 
```

The output:
```
Issue979RemainingErrors.java:29: error: [return.type.incompatible] incompatible types in return.
		return sym;
		       ^
  type of expression: T[ extends @UnknownVal Symbol super @UnknownVal Void]
  method return type: T[ extends @UnknownVal Symbol super @BottomVal Void]
Issue979RemainingErrors.java:37: error: [return.type.incompatible] incompatible types in return.
		return sym;
		       ^
  type of expression: T[ extends @UnknownVal Symbol super @UnknownVal Void]
  method return type: T[ extends @UnknownVal Symbol super @BottomVal Void]
2 errors
```

Note that in each case, the error has some similar properties:
1. the type of the expression is `T extends <Top> super <Top>`
2. the required type is `T extends <Top> super <Bottom>`

Running this example through another type system that should issue no errors, such as the Tainting Checker, shows similar behavior:

```
Issue979RemainingErrors.java:29: error: [return.type.incompatible] incompatible types in return.
		return sym;
		       ^
  type of expression: T[ extends @Tainted Symbol super @Tainted Void]
  method return type: T[ extends @Tainted Symbol super @Untainted Void]
Issue979RemainingErrors.java:37: error: [return.type.incompatible] incompatible types in return.
		return sym;
		       ^
  type of expression: T[ extends @Tainted Symbol super @Tainted Void]
  method return type: T[ extends @Tainted Symbol super @Untainted Void]
2 errors
```

I propose that we implement a check to prevent such errors from being issued, based on the pattern of the required type being `<GenericType> extends <Top> super <Bottom>` when the actual type is `<GenericType> extends <Top> super <Top>`.

I'm happy to do the implementation work, but I'm not exactly sure where this information is available that I'd be able to prevent the error from being issued. I've considered two options:
* in `SourceChecker#report`, but the top and bottom types in the qualifier hierarchy aren't available there (or at least, I don't know how to get to them)
* in the common assignment check, but not all errors go through the CAC. That still might be the best way to do it

@smillst, do you have any other ideas on how I should do this, or on which approach you would take?

@msridhar suggested that I do this in public so he could see our discussion rather than just sending an email, so I'm pinging him

Ok, I retried on a large object construction benchmark and I'm still seeing this issue.  Here is one message that I think matches the `super` pattern @kelloggm mentioned:

```
/Users/msridhar/git-repos/dljc-experiment/outdir/nomulus/core/src/main/java/google/registry/beam/invoicing/InvoicingPipeline.java:154: error: [argument.type.incompatible] incompatible types in argument.
                  .via(BillingEvent::getInvoiceGroupingKey))
                      ^
  found   : @CalledMethodsTop MapElements</*INFERENCE FAILED for:*/ ? extends @CalledMethodsTop Object, /*INFERENCE FAILED for:*/ ? extends @CalledMethodsTop Object>
  required: @CalledMethodsTop PTransform<? extends @CalledMethodsTop PInput super @CalledMethodsTop PCollection<@CalledMethodsTop BillingEvent>, /*INFERENCE FAILED for:*/ @CalledMethodsTop ? extends @CalledMethodsTop POutput>
```

[Here](https://github.com/msridhar/nomulus/blob/base/core/src/main/java/google/registry/beam/invoicing/InvoicingPipeline.java#L154) is the relevant line (I haven't fully understood what is going on).

Here's another one:

```
/Users/msridhar/git-repos/dljc-experiment/outdir/nomulus/core/src/main/java/google/registry/tools/GenerateDnsReportCommand.java:105: error: [return.type.incompatible] incompatible types in return.
                      ImmutableMap.of(
                                     ^
  type of expression: @CalledMethodsTop ImmutableMap<@CalledMethodsTop String, @CalledMethodsTop Serializable & @CalledMethodsTop Comparable<?[ extends @CalledMethodsTop Serializable & @CalledMethodsTop Comparable<?[ extends @CalledMethodsTop Object super @CalledMethodsTop Void]> super @CalledMethodsTop Void]>>
  method return type: @CalledMethodsTop ImmutableMap<@CalledMethodsTop String, @CalledMethodsTop Serializable & @CalledMethodsTop Comparable<?[ extends @CalledMethodsTop Serializable & @CalledMethodsTop Comparable<?[ extends @CalledMethodsTop Object super @CalledMethodsBottom Void]> super @CalledMethodsBottom Void]>>
```

[Relevant line](https://github.com/msridhar/nomulus/blob/base/core/src/main/java/google/registry/tools/GenerateDnsReportCommand.java#L105)

If you need a repro test case let me know, but that will take some time to make.

--- a/framework/src/main/java/org/checkerframework/framework/util/typeinference/DefaultTypeArgumentInference.java
+++ b/framework/src/main/java/org/checkerframework/framework/util/typeinference/DefaultTypeArgumentInference.java
@@ -747,8 +747,7 @@ public class DefaultTypeArgumentInference implements TypeArgumentInference {
             final TypeVariable typeVar = atv.getUnderlyingType();
             if (targets.contains((TypeVariable) TypeAnnotationUtils.unannotatedType(typeVar))) {
                 final AnnotatedTypeMirror inferredType = inferredArgs.get(typeVar);
-                if (inferredType == null
-                        || TypeArgInferenceUtil.containsTypeParameter(inferredType, targets)) {
+                if (inferredType == null) {
                     AnnotatedTypeMirror dummy = typeFactory.getUninferredWildcardType(atv);
                     inferredArgs.put(atv.getUnderlyingType(), dummy);
                 }

--- a/framework/tests/value/Issue3001.java
+++ b/framework/tests/value/Issue3001.java
@@ -1,0 +1,7 @@
+class Issue3001 {
+
+    private <T> T getMember(Class<T> type) {
+        T sym = getMember(type);
+        return sym;
+    }
+}

